Add support for dynamic mappings

advisoray opened the pull request:
Hey,

Here's the code that makes SpaceCadet able to support dynamic key maps.  I figured since it was no longer limited to shift, we could kick the 'Shift' part of the name, especially considering that the new name already matches the repo name.

I updated the example and the readme, too.  Please let me know if you need anything to be changed, or if there are any stylistic or implementation details you'd like me to address.

Thanks,
Ben

algernon commented on examples/SpaceCadet/SpaceCadet.ino:
```diff
@@ -40,7 +40,22 @@ const Key keymaps[][ROWS][COLS] PROGMEM = {
 void setup() {
-  Kaleidoscope.use(&SpaceCadetShift);
+  //Tell Kaleidoscope to use SpaceCadet
+  Kaleidoscope.use(&SpaceCadet;
```
This has a syntax error, the closing `)` is missing. Can you fix that?

algernon commented on src/Kaleidoscope/SpaceCadet.cpp:
```diff
@@ -14,84 +14,176 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Modified by Ben Gemperline to support additional keys and dynamic
+ * key mappings.
```
Just add yourself to the Copyright line! The history should be in git logs, and the copyright line at the top would then reflect who added significant amount of code to the plugin.

algernon commented on src/Kaleidoscope/SpaceCadet.h:
```diff
@@ -21,22 +21,47 @@
 
-  void begin(void) final;
+            //Constructor with mapping
+            SpaceCadet(ModifierKeyMap * map, uint8_t map_size);
```
I think there is little use for this constructor, because the plugin will be used through the `SpaceCadet` object, and mappings will be set with the `setMap` function below.

The only plugins that have ~non-empty constructors are those that do not provide a global object, and *always* require the user to instantiate them. If a plugin provides an instance, then creating another is rarely useful, if at all.

In this case, that wouldn't work as expected, either, because the map is static, thus, shared between all instances.

algernon commented on src/Kaleidoscope/SpaceCadet.h:
```diff
@@ -21,22 +21,47 @@
-};
+        private:
+            static uint8_t map_size;
+            static ModifierKeyMap * map;
```
Can you rename this to `static ModifierKeyMap *map_`, and the one above to `static uint8_t map_size_`?

We did not convert all plugins to our coding style guide, but `SpaceCadet` was one of those that were, and said guide says private instance variables should have a trailing underscore in their name. (Apologies for not referencing the style guide, I have very little time at the moment)

algernon commented:
Whee!

Thanks for the pull request! I like where this is going. Had a quick look at it, and you can see a few comments above. There may be more coming, when I have time to have a more in-depth review, but the general direction I like.

I will try to review it in a timely manner, but can't promise anything right now (becoming a father has dire consequences on one's available time :D).

advisoray commented:
Thanks for all of the helpful comments! I will address this all and get it
updated. Then you can review again at your leisure. No worries on time; I
just wanted to get my code changes back up and public.

On Aug 23, 2017 12:21 AM, "Gergely Nagy" <notifications@github.com> wrote:

> Whee!
>
> Thanks for the pull request! I like where this is going. Had a quick look
> at it, and you can see a few comments above. There may be more coming, when
> I have time to have a more in-depth review, but the general direction I
> like.
>
> I will try to review it in a timely manner, but can't promise anything
> right now (becoming a father has dire consequences on one's available time
> :D).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/keyboardio/Kaleidoscope-SpaceCadet/pull/4#issuecomment-324232862>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AdiDXAC70u9AaokV7M8hAEv88RPvD6O1ks5sa8TsgaJpZM4O_eoO>
> .
>

advisoray commented:
I updated the code to address the style guide changes and fixed the example so it will compile.  I also ran astyle and fixed the indentation / style for that, too, so the travis-CI build works again, as well.  Please let me know if there's anything else you'd like to see done!

- Ben

advisoray commented:
Just wanted to check in and see if there's anything else you'd like me to do.

obra commented:
I suspect Algernon just needs a few more hours in the day. I'm not really
clueful enough about this code to review it and he has twin newborns.

On Fri, Sep 1, 2017 at 8:51 PM, Ben Gemperline <notifications@github.com>
wrote:

> Just wanted to check in and see if there's anything else you'd like me to
> do.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/keyboardio/Kaleidoscope-SpaceCadet/pull/4#issuecomment-326711094>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACxaLmYbOO-kEOq3_gbpgQttWsN1U7xks5seKaTgaJpZM4O_eoO>
> .
>

advisoray commented:
Haha, thanks for the update! No worries. Just trying to make sure it didn't
get lost, is all.

On Sep 1, 2017 7:06 PM, "Jesse Vincent" <notifications@github.com> wrote:

> I suspect Algernon just needs a few more hours in the day. I'm not really
> clueful enough about this code to review it and he has twin newborns.
>
> On Fri, Sep 1, 2017 at 8:51 PM, Ben Gemperline <notifications@github.com>
> wrote:
>
> > Just wanted to check in and see if there's anything else you'd like me to
> > do.
> >
> > —
> > You are receiving this because you are subscribed to this thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/keyboardio/Kaleidoscope-SpaceCadet/pull/4#
> issuecomment-326711094>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/AACxaLmYbOO-kEOq3_
> gbpgQttWsN1U7xks5seKaTgaJpZM4O_eoO>
> > .
> >
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/keyboardio/Kaleidoscope-SpaceCadet/pull/4#issuecomment-326712065>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AdiDXJVn3nwAH91W-BdSc2o2_lrb3r-Pks5seKoDgaJpZM4O_eoO>
> .
>

algernon commented:
No, it's not lost, I just don't have the time to review your PR at the moment. Give me a few more days, and I'll get around to it, it's near the top of my TODO list.

advisoray commented:
Yep, completely understandable! Good luck with the newborns! They are much
more important than this. 😁

On Sep 1, 2017 10:52 PM, "Gergely Nagy" <notifications@github.com> wrote:

> No, it's not lost, I just don't have the time to review your PR at the
> moment. Give me a few more days, and I'll get around to it, it's near the
> top of my TODO list.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/keyboardio/Kaleidoscope-SpaceCadet/pull/4#issuecomment-326722020>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AdiDXLxeIuByzgTm0FxX38vXlETAZ1SRks5seN8QgaJpZM4O_eoO>
> .
>

algernon commented on README.md:
```diff
@@ -25,6 +25,15 @@ and it still would act as a `Shift`, without the parens inserted on release:
+plugin is configurable and extensible to support adding symbols to other keys,
+as well.  Along with `(` on your left `Shift` key, you with to add other such
+programming mainstays as `{` to your `cmd` key and `[` to your left `Control`
+key (and, of course, the matching bookend symbols on the right).  You can do
```
This last sentence is a bit unclear: `you with to add other such...`? I assume that `with` should have been `wish`?

advisoray replied:
Yep, definitely. 'wish'.

algernon commented on README.md:
```diff
@@ -25,6 +25,15 @@ and it still would act as a `Shift`, without the parens inserted on release:
+this sort of functionality on other keys, as well.  Fortunately, the Space Cadet
+plugin is configurable and extensible to support adding symbols to other keys,
+as well.  Along with `(` on your left `Shift` key, you with to add other such
+programming mainstays as `{` to your `cmd` key and `[` to your left `Control`
```
`cmd` => `Cmd`/`Alt`, to be less mac-specific, perhaps

advisoray replied:
I was just going by the actual labeling of the keyboard.  The left-side key where `alt` would be is labeled as `cmd`, at least on the PVT board I have.  That's my reasoning, anyways.  Having Cmd/Alt is actually more confusing, to me, because I was talking about the left side bindings only.  I could simply change it to talk about the right side bindings, too, and get rid of the confusion (the approach I will likely do).

algernon commented on README.md:
```diff
@@ -37,38 +46,74 @@ enabling the plugin:
+    ,{Key_RightControl, Key_RightBracket, 250}
+  };
+  //Tell SpaceCadet to use the map  
+  SpaceCadet.setMap(spacecadetmap, sizeof(spacecadetmap)/sizeof(spacecadetmap[0]));
```
The `sizeof(...)/sizeof(...[0])` idiom is... not very friendly. Can we have a macro that does this for the user? Something like this:

```c++
#define SPACECADET_MAP(map) map, sizeof(map)/sizeof(map[0])

SpaceCadet.setMap(SPACECADET_MAP(spacecadetmap));
```

advisoray replied:
I agree that it is kludgy, and that was actually the thing I felt least happy with.  I think I'll go with the NoKey/NoKey termination (though I don't really like that, either, but it's the less-kludgy option) and update the setMap function.  I was also thinking of maybe making it a Vector instead, but don't know how much extra useless stuff we'd include by including the std library (and memory is at a premium).

algernon replied:
`std::vector` would add waaay too much bloat, as you'd need to pull in an entire standard c++ library (Arduino does not ship with one, it only has a very small subset of a c++ library).

A third option would be to use a function, which the user can override, a'la `macroAction`, `tapDanceAction` and the like:

```c++
kaleidoscope::SpaceCadet::something spaceCadetAction(Key key) {
  switch (key.raw) {
    case Key_LeftShift.raw:
        return {Key_LeftParen, 250};
  }
  return {Key_NoKey, 0};
}
```

This saves us a sentinel, but... it's even more awkward to use.

advisoray replied:
I ended up implementing the sentinel, and further implemented the SPACECADET_MAP_END macro to represent it, so it's as easy as setting the map.  I got rid of the setMap method, too.

algernon commented on examples/SpaceCadet/SpaceCadet.ino:
```diff
@@ -40,7 +40,22 @@ const Key keymaps[][ROWS][COLS] PROGMEM = {
+    , {Key_RightControl, Key_RightBracket, 250}
+  };
+  //Set the map and the number of entries
+  SpaceCadet.setMap(spacecadetmap, sizeof(spacecadetmap) / sizeof(spacecadetmap[0]));
```
See the README comment about the macro - when we have such a macro, please use it here too.

advisoray replied:
Updated the readme and the example.

algernon commented on src/Kaleidoscope/SpaceCadet.cpp:
```diff
@@ -21,35 +21,83 @@
+  input = input_;
+  output = output_;
+  timeout = timeout_;
+}
```
I wonder if it would make more sense to move this class under `SpaceCadet`? It is the only user of it, anyway. I'm also torn about the name... a single instance of this class is not a map, it is an element of a map.

How about (after moving it under `kaleidoscope::SpaceCadet`) `kaleidoscope::SpaceCadet::KeyBinding`? Not entirely sold on that, either... but `ModifierKeyMap` sounds wrong.

A small thing, but - I think - an important one.

advisoray replied:
Yeah, when I started it it wasn't actually a single element, but the whole map.  I'll move it under and call it KeyBinding.  I wasn't quite sure what name to give it, either, but KeyBinding sounds the most like what it truly is, to me.

advisoray replied:
Moved to `kaleidoscope::SpaceCadet::KeyBinding`

algernon commented on src/Kaleidoscope/SpaceCadet.cpp:
```diff
@@ -21,35 +21,83 @@
+
+  //Set the variables to our internal map
+  map_ = internalMap;
+  map_size_ = sizeof(internalMap) / sizeof(internalMap[0]);
```
Would it be possible to initialize `map_` and `map_size_` outside of the constructor? Haven't checked if it works, but:

```c++
ModifierKeyMap *SpaceCadet::map_ = {
  {Key_LeftShift, Key_LeftParen, 0},
  {Key_RightShift, Key_RightParen, 0}
};
uint8_t SpaceCadet::map_size_ = sizeof(SpaceCadet::map_)/sizeof(ModifierKeyMap);
```

My gut feeling is that this would result in slightly smaller code. But then, I haven't tried compiling it, so YMMV.

advisoray replied:
I'll take a look and see.

advisoray replied:
While you could probably get it to work, it's not so easy.  I want the map to be a pointer to a map so that it can be arbitrarily sized. I tried playing around with initializing it outside of the constructor, and I didn't get anywhere on that front.  I could do it if I made it a fixed-size map, but that doesn't help if you want a differently-sized map.  In the end, I think simply having the empty constructor and setting it will work the best.  If you can get it to work properly without the empty constructor, please update it and let me see how you are able to do it.  `C++` is fairly new to me, though I have a bunch of experience in `C` and `Java`, which both do things slightly differently.

algernon commented on src/Kaleidoscope/SpaceCadet.cpp:
```diff
@@ -21,35 +21,83 @@
+  map_ = map;
+  //set the map size to be the length of the array
+  map_size_ = map_size;
+}
```
I was thinking... what if we required the map to end with a sentinel? (`{Key_NoKey, Key_NoKey, 0}`, hidden behind a well-named macro, say, `SPACECADET_MAP_END`)

If we did that, we didn't need the `setMap` method, and could just use `SpaceCadet.map = my_own_map`.

What do you think?

advisoray replied:
I like it.  I like the Macro even better.  I'll implement it this way.

advisoray replied:
Implemented as such.

algernon commented on src/Kaleidoscope/SpaceCadet.cpp:
```diff
@@ -58,40 +106,75 @@ Key SpaceCadetShift::eventHandlerHook(Key mapped_key, byte row, byte col, uint8_
 
-kaleidoscope::SpaceCadetShift SpaceCadetShift;
+//Note: since this is more generic, it is now simply SpaceCadet, instead of
+//SpaceCadetShift
```
I think we can leave out the reference to the old name here. As the plugin did not hit 1.0 yet, API changes are to be expected.

Until MP units start to ship, I wouldn't worry about breaking things a little.

advisoray replied:
Got it.

advisoray commented:
All changes requested have been made or explained.

obra commented:
Merging, per @algernon's +1

Thanks so much for the patches, Ben!
